### PR TITLE
feat(vercel): ISR

### PR DESCRIPTION
**Pull request by lilnasy**
## Changes

- Introduces a project level config that enables persistent caching for all on-demand rendered routes.
- [x] website-wide isr
- [x] website-wide revalidate
- [x] website-wide bypassToken
- [x] exclude paths from isr
- [ ] ~~route-specific configuration~~ out of scope
- [x] dev mode should not allow query params when isr is enabled

## Usage

<details>
<summary>Basic usage - enough for 80% of the users - caches all pages on first request and saves forever</summary>

```ts
export default defineConfig({
    output: "server",
    adapter: vercel({ isr: true })
})
```

</details>

<details>
<summary>Time based invalidation - caches all pages on first request and saves for 1 day</summary>

```ts
export default defineConfig({
    output: "server",
    adapter: vercel({
        isr: {
            expiration: 60 * 60 * 24
        }
    })
})
```

</details>

<details>
<summary>Manual invalidation - caches all pages on first request and lets you create an endpoint where you clear it</summary>

```ts
export default defineConfig({
    output: "server",
    adapter: vercel({
        isr: {
            // A random string that you create. Its presence in the `__prerender_bypass` cookie will result in fresh responses being served, bypassing the cache.
            bypassToken: "005556d774a8" // I created this using crypto.randomUUID(),
            // Paths that will always be served fresh.
            exclude: [ "/api/invalidate" ] 
        }
    })
})
```

</details>

## Testing
Added isr.test.js with a fixture

## Docs
 - withastro/docs#6739

### CODE REVIEW

**Comment by lilnasy** [comment]
!preview isr

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (natemoo-re) [inline packages/integrations/vercel/src/serverless/adapter.ts:143]: I assume this value should always remain secret? Do we want to clarify that in the JSDoc comment?

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (natemoo-re) [inline packages/integrations/vercel/src/serverless/adapter.ts:271]: This seems good, but I worry that this will be confusing because it happens silently. Maybe log a warning?

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (natemoo-re) [inline packages/integrations/vercel/src/serverless/adapter.ts:354]: Avoid errors when `isr === null` since `typeof null === 'object'`. Technically not a valid type, but I've been bitten enough to watch for this 😓 

```suggestion
					if (isr === false || (isr && typeof isr === "object" && isr.exclude)) {
```

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (natemoo-re) [inline packages/integrations/vercel/src/serverless/adapter.ts:368]: Same
```suggestion
					if (isr && typeof isr === "object" && isr.exclude) {
```

**Review by natemoo-re — APPROVED** [review]
Great work! I'd love at least one other approval before we merge this given what a big feature this is!

**Review by natemoo-re — APPROVED** [review]
Great work! I'd love at least one other approval before we merge this given what a big feature this is!

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts:271]: Yeah, I agree. Do we already have a message for when `astro dev` removes params?

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [inline packages/integrations/vercel/src/serverless/adapter.ts:142]:
```suggestion
	/**
	 * A secret random string that you create.
	 * Its presence in the `__prerender_bypass` cookie will result in fresh responses being served, bypassing the cache.
	 * Its presence in the `x-prerender-revalidate` header will result in a fresh response which will then be cached for all future requests to be used.
	 * 
	 * By default, none.
	 */
```

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (natemoo-re) [reply packages/integrations/vercel/src/serverless/adapter.ts:272]: Ha, no. Turns out that we do it silently in core as well. https://github.com/withastro/astro/blob/7224809b73d2c3ec8e8aee2fa07463dc3b57a7a2/packages/astro/src/vite-plugin-astro-server/request.ts#L52

I retract my suggestion!

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (bluwy) [inline packages/integrations/vercel/src/serverless/adapter.ts:355]: I'm a bit confused about this check, should this be an `else` block because we already check `isr` above?

I'm not sure why this is checking `isr.exclude` specifically to create another function folder which could've been already generated above.

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (ematipico) [inline packages/integrations/vercel/src/serverless/adapter.ts:136]: I don't see the point of having a `VercelISRConfig` config where all the fields are optional. This means that a user can do:

```js
{
	isr: {}
}
```

Which feels wrong

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (ematipico) [reply packages/integrations/vercel/src/serverless/adapter.ts:266]: +1 from me too. Also, if a user passes a string, we should do an equality check, and not create a regex.

Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts:266]: I agree we should be stricter.

It's a RegExp here because that's how vercel processes the routes file.

I'll look into making the exclude paths behave as literal paths.

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (ematipico) [inline packages/integrations/vercel/src/serverless/adapter.ts:369]: We keep doing the same checks, can't we create an internal isr object only once from the options of the user?

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts:355]: The condition here checks whether a normal function needs to be created. Even when ISR is enabled, the excluded paths would need to create one.

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (ematipico) [inline packages/integrations/vercel/src/serverless/adapter.ts:528]: The extraction of the fields here can be cleaner, especially considering that we do `typeof isr === "object"` every single time we need to read `isr`

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts]: Cleaned up. [Here is the new control flow](https://github.com/withastro/astro/pull/9714/files#diff-071ccfbdfd7b9bc803219179f88ae0bdb227a399bc4a481b7831c82b801d7a17R332-R344)

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts:149]: The default javascript value is false, but that just doesn't communicate anything.

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts:138]: Admittedly, the descriptions here are brief, but that's intentional. We really are only copy-pasting values from one file to another. To really understand the concepts, we kinda need to defer to vercel docs.

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts]: Cleaned up. [Here is the new control flow](https://github.com/withastro/astro/pull/9714/files#diff-071ccfbdfd7b9bc803219179f88ae0bdb227a399bc4a481b7831c82b801d7a17R332-R344)

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts]: Cleaned up. [Here is the new control flow](https://github.com/withastro/astro/pull/9714/files#diff-071ccfbdfd7b9bc803219179f88ae0bdb227a399bc4a481b7831c82b801d7a17R487-R497)

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (ematipico) [reply packages/integrations/vercel/src/serverless/adapter.ts:138]: I see, then I suppose it makes total sense to add a link to their docs :)

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (ematipico) [reply packages/integrations/vercel/src/serverless/adapter.ts:149]: That's false because the parent object is optional, right? But if I pass an object, this field must have a default as stated in the docs.

Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts:149]: Does this help?

image

[prerender-config.json](https://vercel.com/docs/build-output-api/v3/primitives#prerender-configuration-file)

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts:136]: That is a possibility, but each one of the options really is optional and can individually be used. `isr: {}` would effectively be `isr: true`.

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts:138]: Updated lines [137-144](https://github.com/withastro/astro/pull/9714/files/#diff-071ccfbdfd7b9bc803219179f88ae0bdb227a399bc4a481b7831c82b801d7a17R137-R144)

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (lilnasy) [reply packages/integrations/vercel/src/serverless/adapter.ts]: New lines [146-153](https://github.com/withastro/astro/pull/9714/files#diff-071ccfbdfd7b9bc803219179f88ae0bdb227a399bc4a481b7831c82b801d7a17R146-R153).

**Comment by lilnasy** [comment]
Converting to draft because there's something about routing I need to confirm after these changes, but I can't today. Reviews can continue.

File: packages/integrations/vercel/src/serverless/adapter.ts
Comment (ematipico) [reply packages/integrations/vercel/src/serverless/adapter.ts:136]: JavaScript, what a language... 😔

**Comment by sarah11918** [comment]
I was asked to review the typed documentation here, and it's looking pretty good! Right now, there's nothing here that would show up in docs itself, nor a changeset to edit. But, I'd say this is a good basis for API docs. Carry on! 🫡

**Comment by lilnasy** [comment]
Oh, didn't even realise I missed the changeset.

File: .changeset/flat-snakes-hammer.md
Comment (lilnasy) [inline .changeset/flat-snakes-hammer.md:26]:
```suggestion
            expiration: 60 * 60 * 24
```

File: .changeset/flat-snakes-hammer.md
Comment (sarah11918) [inline .changeset/flat-snakes-hammer.md:5]:
```suggestion
Introduces a new config option, `isr`, that allows you to deploy your project as an ISR function. [ISR (Incremental Static Regeneration)](https://vercel.com/docs/incremental-static-regeneration) caches your on-demand rendered pages in the same way as prerendered pages after first request.
```

File: .changeset/flat-snakes-hammer.md
Comment (sarah11918) [inline .changeset/flat-snakes-hammer.md:7]:
```suggestion
To enable this feature, set `isr` to true in your Vercel adapter configuration in `astro.config.mjs`:
```
Even though this is an *adapter's* minor, and not an Astro one, I still do kind of want to kind of emphasize that this is a specific adapter setting, and unique to the Vercel adapter, to avoid any confusion.

File: .changeset/flat-snakes-hammer.md
Comment (sarah11918) [inline .changeset/flat-snakes-hammer.md:48]:
``````suggestion

## Cache invalidation options

By default, ISR responses are cached for the duration of your deployment. You can further control caching by setting an `expiration` time or prevent caching entirely for certain routes.

### Time-based invalidation

You can change the length of time to cache routes this by configuring an `expiration` value in seconds:

```js
export default defineConfig({
    output: "server",
    adapter: vercel({
        isr: {
            // caches all pages on first request and saves for 1 day
            expiration: 60 * 60 * 24
        }
    })
})
```

### Manual invalidation

To implement Vercel's [Draft mode](https://vercel.com/docs/build-output-api/v3/features#draft-mode), or [On-Demand Incremental Static Regeneration (ISR)](https://vercel.com/docs/build-output-api/v3/features#on-demand-incremental-static-regeneration-isr), you can create a bypass token and provide it to the `isr` config along with the paths to exclude from caching: 

```js
export default defineConfig({
    output: "server",
    adapter: vercel({
        isr: {
            // A secret random string that you create.
            bypassToken: "005556d774a8",
            // Paths that will always be served fresh.
            exclude: [ "/api/invalidate" ] 
        }
    })
})
```
`````` 
I'm not sure we can just jump to random h5s out of the blue like that.  :smile: 

Some other subtle suggestions for tweaks:

- Create one header for these two options, that are both ways of controling/avoiding caches
- Put the default behaviour *there*, in the preamble, since it's appropriate for both (and good to state a default upfront)
- Be explicit about `expiration` being in seconds
- (gut feeling?) Since we are linking to *Vercel's* docs, and using their terms, call these things "Vercel's Draft mode.. etc." Especially since this is the Vercel adapter, I think it's perfectly fine to take some of the accountability off of us, and make it clear "Our Vercel adapter can handle these Vercel features" and make it clear this is not an "Astro" thing, just a compatibility with/implementation of Vercel thing.

File: .changeset/flat-snakes-hammer.md
Comment (sarah11918) [inline .changeset/flat-snakes-hammer.md:50]:
```suggestion
Note that ISR function requests do not include search params, similar to [requests](https://docs.astro.build/en/reference/api-reference/#astrorequest) in static mode.
```

I'm torn about whether this really belongs here in the changelog statement, vs a "Read more in our docs" link. (Which we absolutely should include here)

**Review by sarah11918 — COMMENTED** [review]
Thanks @lilnasy ! Looking good! Here are my initial thoughts!

File: .changeset/flat-snakes-hammer.md
Comment (lilnasy) [reply .changeset/flat-snakes-hammer.md:55]: We can remove it entirely. It is vercel's behavior not ours.

**Pull request merged**
